### ISSUE TITLE
Passing metadata to service implementation

### ISSUE THREAD

**Issue by charleskorn** [square/wire#1930]
I'd like to be able to pass incoming gRPC metadata to a service implementation method - so a request handler can get access to the metadata as part of handling that request. My personal preference would be that it's passed as a parameter to the method, but I'm open to other ideas as well.

Using the sample app as an example, I'd like implementation methods like [`WhiteboardGrpcAction.Whiteboard()`](https://github.com/square/wire/blob/cbbe3fe5f25f25808c9e739d1754c5339e705135/wire-grpc-sample/server/src/main/java/com/squareup/wire/whiteboard/WhiteboardGrpcAction.kt#L30) to include the metadata as a parameter:

```diff
   override fun Whiteboard(
     request: MessageSource<WhiteboardCommand>,
-    response: MessageSink<WhiteboardUpdate>
+    response: MessageSink<WhiteboardUpdate>,
+    metadata: Metadata 
   ) {
```

Is this something that is already implemented or could easily be added?

**Comment by swankjesse** [comment]
Have you considered passing metadata via the class constructor? In Misk we do ActionScoped for this, but we could also pass direct values to constructors; using instances for one call only.

**Comment by swankjesse** [comment]
(100% agree this is a problem, and I’d like a solution that’ll be easy to do in arbitrary backends).

**Comment by charleskorn** [comment]
> Have you considered passing metadata via the class constructor? In Misk we do ActionScoped for this, but we could also pass direct values to constructors; using instances for one call only.

Yeah that was my other idea. Passing the metadata as a parameter works better with my current setup, as the service instances are currently not request-scoped, but creating a new instance and passing the metadata to the constructor would be possible as well. In an ideal world, it'd be great to be able to choose between the two styles depending on your situation.

**Comment by swankjesse** [comment]
What server are you using? One thing that's nice with the current structure is the server binding is very regular. If we introduce more parameters that might also need extra annotations etc.

**Comment by charleskorn** [comment]
I'm implementing a client for Docker's BuildKit, so I have a little dispatcher I've put together, rather than a server. Building a Docker image using BuildKit's API requires making a HTTP/1 request to an endpoint on the Docker daemon that then hijacks the underlying TCP connection and uses that for gRPC requests over HTTP/2 back to the client.
